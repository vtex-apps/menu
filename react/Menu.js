import React, { Component } from 'react'
import PropTypes from 'prop-types'
import VTEXClasses from './CSSClasses'
import { intlShape, injectIntl } from 'react-intl'

import { Link } from 'render'

import './global.css'

const MAX_ITEMS = 10

/**
 * Links Menu Component. Shows a menu bar with links.
 */
class Menu extends Component {
  constructor(props) {
    super(props)
    MenuWithIntl.intl = props.intl
    MenuWithIntl.uiSchema = MenuWithIntl.getUiSchema()
  }

  static defaultProps = {
    numberOfItems: 0,
  }

  static propTypes = {
    /** Number of items. */
    numberOfItems: PropTypes.number.isRequired,
    /** Intl instance. */
    intl: intlShape.isRequired,
  }

  componentWillReceiveProps(props) {
    if (this.props.intl !== props.intl) {
      MenuWithIntl.intl = props.intl
      MenuWithIntl.uiSchema = MenuWithIntl.getUiSchema()
    }
  }

  renderLink(link) {
    let className = 'f6 link dib white dim mr3 mr4-ns'
    switch (link.position) {
      case 'LEFT':
        className = `${VTEXClasses.LINK_LEFT} ${className}`
        break
      case 'MIDDLE':
        className = `${VTEXClasses.LINK_MIDDLE} ${className}`
        break
      case 'RIGHT':
        className = `${VTEXClasses.LINK_RIGHT} ${className}`
        break
    }
    return (
      <Link key={link.title} className={`${className}`} page={link.url}>
        {link.title}
      </Link>
    )
  }

  getLinksFromProps() {
    const links = []
    for (let i = 1; i <= this.props.numberOfItems; i++) {
      this.props[`item${i}`] && links.push(this.props[`item${i}`])
    }
    return links
  }

  render() {
    const links = this.getLinksFromProps()
    return (
      <div className={`${VTEXClasses.MAIN_CLASS} w-100 dn db-ns`}>
        <nav className="flex justify-between bb b--white-10 bg-near-black">
          <div className="flex-grow pa3 flex items-center">
            {links.filter(link => link['position'] === 'LEFT').map(link => {
              return this.renderLink(link)
            })}
          </div>
          <div className="flex-grow pa3 flex items-center">
            {links.filter(link => link['position'] === 'MIDDLE').map(link => {
              return this.renderLink(link)
            })}
          </div>
          <div className="flex-grow pa3 flex items-center">
            {links.filter(link => link['position'] === 'RIGHT').map(link => {
              return this.renderLink(link)
            })}
          </div>
        </nav>
      </div>
    )
  }
}

const MenuWithIntl = injectIntl(Menu)

MenuWithIntl.getUiSchema = () => {
  const intl = MenuWithIntl.intl
  const titleIntl = (intl && intl.formatMessage({ id: 'menu.title' })) || 'Title'

  const uiSchema = {
    numberOfItems: {
      'ui:widget': 'range',
    },
  }
  for (let i = 1; i <= MAX_ITEMS; i++) {
    uiSchema[`item${i}`] = {
      title: { 'ui:placeholder': titleIntl },
      url: { 'ui:placeholder': 'URL' },
      position: {
        'ui:widget': 'radio',
        'ui:options': { 'inline': true },
      },
    }
  }
  return uiSchema
}

MenuWithIntl.getSchema = ({ numberOfItems }) => {
  const dynamicProperties = {}

  const intl = MenuWithIntl.intl
  const descriptionIntl = (intl && intl.formatMessage({ id: 'menu.description' })) || 'A menu bar of links'
  const titleIntl = (intl && intl.formatMessage({ id: 'menu.title' })) || 'Title'
  const numberOfItemsIntl = (intl && intl.formatMessage({ id: 'menu.numberOfItems' })) || 'Number of items'
  const positionIntl = (intl && intl.formatMessage({ id: 'menu.position' })) || 'Position'
  const leftIntl = (intl && intl.formatMessage({ id: 'menu.position.left' })) || 'LEFT'
  const middleIntl = (intl && intl.formatMessage({ id: 'menu.position.middle' })) || 'MIDDLE'
  const rightIntl = (intl && intl.formatMessage({ id: 'menu.position.right' })) || 'RIGHT'

  for (let i = 1; i <= numberOfItems; i++) {
    dynamicProperties[`item${i}`] = {
      type: 'object',
      title: `Item #${i}`,
      required: ['title', 'url', 'position'],
      properties: {
        title: {
          title: titleIntl,
          type: 'string',
        },
        url: {
          title: 'URL',
          type: 'string',
        },
        position: {
          title: positionIntl,
          type: 'string',
          enum: ['LEFT', 'MIDDLE', 'RIGHT'],
          enumNames: [leftIntl, middleIntl, rightIntl],
          default: 'MIDDLE',
        },
      },
    }
  }
  const schema = {
    title: 'Menu',
    description: descriptionIntl,
    type: 'object',
    properties: {
      numberOfItems: {
        title: numberOfItemsIntl,
        type: 'number',
        default: 0,
        minimum: 0,
        maximum: MAX_ITEMS,
      },
    },
  }

  schema.properties = {
    ...schema.properties,
    ...dynamicProperties,
  }

  return schema
}

export default MenuWithIntl
