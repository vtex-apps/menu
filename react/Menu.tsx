import React, { useContext, useMemo } from 'react'

import classNames from 'classnames'
import { defineMessages } from 'react-intl'

import { generateBlockClass } from '@vtex/css-handles'
import CategoryMenu from './components/CategoryMenu'
import Item from './components/Item'
import LevelContext from './components/LevelContext'
import MenuContext from './components/MenuContext'
import MenuItem, { MenuItemSchema } from './MenuItem'

import styles from './Menu.css'

const TypographyMap: Record<string, string> = {
  body: 't-body',
  heading1: 't-heading-1',
  heading2: 't-heading-2',
  heading3: 't-heading-3',
  heading4: 't-heading-4',
  heading5: 't-heading-5',
  mini: 't-mini',
  small: 't-small',
}

const Menu: StorefrontFunctionComponent<MenuSchema> = ({
  orientation = 'horizontal',
  textType,
  title,
  categoryId,
  blockClass,
  items: itemsProps = [],
  children,
}) => {
  const level = useContext(LevelContext)
  const menuContext = useMemo(
    () => ({
      hasTitle: title || categoryId ? true : false,
      orientation,
      textType: textType ? TypographyMap[textType] : TypographyMap.body,
    }),
    [orientation, textType]
  )

  const menuItems = itemsProps.map(
    ({ itemProps: { text }, itemProps, ...rest }) => (
      <MenuItem key={text} itemProps={itemProps} {...rest} />
    )
  )

  const classes = generateBlockClass(styles.menuContainer, blockClass)

  return (
    <LevelContext.Provider value={level + 1}>
      <MenuContext.Provider value={menuContext}>
          <nav>
            <ul
              className={classNames(classes, 'list flex pl0 mv0', {
                'flex-column': orientation === 'vertical',
                'flex-row': orientation === 'horizontal',
              })}
            >
              {!categoryId && title && <Item {...title} isTitle />}
              {categoryId && <CategoryMenu categoryId={categoryId} />}
              {children}
              {menuItems}
            </ul>
          </nav>
      </MenuContext.Provider>
    </LevelContext.Provider>
  )
}

interface MenuSchema {
  orientation?: 'vertical' | 'horizontal'
  categoryId?: number
  textType?: Typography
  title?: MenuItemSchema
  additionalDef?: string
  blockClass?: string
  items?: MenuItemSchema[]
}

enum Typography {
  heading1 = 't-heading-1',
  heading2 = 't-heading-2',
  heading3 = 't-heading-3',
  heading4 = 't-heading-4',
  heading5 = 't-heading-5',
  body = 't-body',
  small = 't-small',
  mini = 't-mini',
}

const messages = defineMessages({
  categoryDef: {
    defaultMessage: '',
    id: 'admin/editor.menu.def.category',
  },
  categoryIdTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.categoryId.title',
  },
  defTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.additionalDef.title',
  },
  horizontalLabel: {
    defaultMessage: '',
    id: 'admin/editor.menu.orientation.horizontal.label',
  },
  menuTitle: {
    defaultMessage: '',
    id: 'admin/editor.future-menu.title',
  },
  noneDef: {
    defaultMessage: '',
    id: 'admin/editor.menu.def.none',
  },
  orientationTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.orientation.title',
  },
  titleDef: {
    defaultMessage: '',
    id: 'admin/editor.menu.def.title',
  },
  typographyTitle: {
    defaultMessage: '',
    id: 'admin/editor.menu.typography.title',
  },
  verticalLabel: {
    defaultMessage: '',
    id: 'admin/editor.menu.orientation.vertical.label',
  },
})

Menu.getSchema = () => {
  // tslint:disable: object-literal-sort-keys
  return {
    title: messages.menuTitle.id,
  }
}

export default Menu
