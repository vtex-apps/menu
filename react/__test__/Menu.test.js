/* eslint-env jest */
import React from 'react'
import { render } from 'react-testing-library'
import { IntlProvider } from 'react-intl'
import Menu from '../Menu'

describe('Menu Component', () => {
  let wrapper

  beforeEach(() => {
    const messages = require('../locales/en-US')
    wrapper = render(
      <IntlProvider locale="en-US" messages={messages}>
        <Menu />
      </IntlProvider>
    )
  })

  it('should be rendered', () => {
    expect(wrapper).toBeDefined()
  })

  it('should match snapshot', () => {
    expect(wrapper.container).toMatchSnapshot()
  })

  it('should render nothing', () => {
    expect(
      wrapper.container.querySelectorAll('a').length
    ).toBe(0)
  })
})
