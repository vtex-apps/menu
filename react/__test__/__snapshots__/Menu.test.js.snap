// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Component should match snapshot 1`] = `
<div>
  <div
    class="vtex-menu w-100"
  >
    <nav
      class="flex justify-between bb b--white-10 bg-near-black"
    >
      <div
        class="flex-grow pa3 flex items-center"
      />
      <div
        class="flex-grow pa3 flex items-center"
      />
      <div
        class="flex-grow pa3 flex items-center"
      />
    </nav>
  </div>
</div>
`;
